### PR TITLE
Fix tables bugs

### PR DESCRIPTION
**Pull request by schangccg**

### CODE REVIEW

**Comment by schangccg** [comment]
for tickets:
https://github.com/muccg/bpa-archive-ops/issues/579
https://github.com/muccg/bpa-archive-ops/issues/580

File: bpaotu/bpaotu/query.py [hunk 1]
```diff
@@ -173,37 +187,64 @@ def _q_all_cached(self, topic, q, mutate_result=None):
...
+    def matching_sample_headers(self, required_headers=None, sort_col=None, sort_order=None):
         query_headers = [SampleContext.id, SampleContext.environment_id]
+        joins = [] # Keep track of any foreign ontology classes which may be needed to be joined to.
 
         cache_name = ['matching_sample_headers']
         if required_headers:
             cache_name += required_headers
             for h in required_headers:
-                query_headers.append(getattr(SampleContext, h))
+                if h != '':
```
Comment (grahame) [inline bpaotu/bpaotu/query.py]: I generally prefer keeping "skip" clauses like this out of the main flow; so I'd do:

```
if h != '': # (or maybe if not h)?
  continue
...
```

File: bpaotu/bpaotu/query.py [hunk 2]
```diff
@@ -173,37 +187,64 @@ def _q_all_cached(self, topic, q, mutate_result=None):
...
+            cache_name.append(str(query_headers[int(sort_col)]))
+            cache_name.append(sort_order)
+
+        elif sort_order == 'desc':
+            q = self._session.query(*query_headers).outerjoin(*joins).order_by(query_headers[int(sort_col)].desc())
+
+            cache_name.append(str(query_headers[int(sort_col)]))
+            cache_name.append(sort_order)
+        else:
+            q = self._session.query(*query_headers).outerjoin(*joins)
```
Comment (grahame) [inline bpaotu/bpaotu/query.py]: you can probably factor out some of this.

I'd move the common bits to the top:

```
q = self._session.query(*query_headers).outerjoin(*joins)
```

then update q to add the order_by stuff if a sort_order is set.

File: bpaotu/bpaotu/query.py [hunk 3]
```diff
@@ -173,37 +187,64 @@ def _q_all_cached(self, topic, q, mutate_result=None):
...
 
-        q = self._session.query(*query_headers)
         return self._q_all_cached(':'.join(cache_name), q)
 
+
```
Comment (grahame) [inline bpaotu/bpaotu/query.py]: does this flake8?

File: bpaotu/bpaotu/query.py (hunk 1 above)
Comment (schangccg) [reply bpaotu/bpaotu/query.py]: Fixed.

File: bpaotu/bpaotu/query.py (hunk 3 above)
Comment (schangccg) [reply bpaotu/bpaotu/query.py]: Fixed.

File: bpaotu/bpaotu/query.py (hunk 2 above)
Comment (schangccg) [reply bpaotu/bpaotu/query.py]: Fixed.

**Pull request merged**